src/audio/codec_adapter/codec/generic.c

```diff
@@ -0,0 +1,366 @@
+		 (uint32_t)ptr);
+
+	return -EINVAL;
+}
+
+static int validate_config(struct codec_config *cfg)
+{
+	//TODO: validation of codec specific setup config
+	return 0;
+}
+
+int codec_prepare(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint32_t codec_id = cd->ca_config.codec_id;
+	struct codec_data *codec = &cd->codec;
+
+	comp_dbg(dev, "codec_prepare() start");
+
+	if (cd->codec.state == CODEC_PREPARED)
+		return 0;
+	if (cd->codec.state < CODEC_INITIALIZED)
+		return -EPERM;
+
+	ret = codec->ops->prepare(dev);
+	if (ret) {
+		comp_err(dev, "codec_prepare() error %d: codec specific prepare failed, codec_id 0x%x",
+			 ret, codec_id);
+		goto end;
+	}
+
+	codec->s_cfg.avail = false;
+	codec->r_cfg.avail = false;
+	codec->r_cfg.data = NULL;
+	list_init(&codec->memory.mem_list);
+
+	if (codec->r_cfg.data)
+		rfree(codec->r_cfg.data);
+
+	codec->state = CODEC_PREPARED;
+	comp_dbg(dev, "codec_prepare() done");
+end:
+	return ret;
+}
+
+int codec_process(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint32_t codec_id = cd->ca_config.codec_id;
+	struct codec_data *codec = &cd->codec;
+
+	comp_dbg(dev, "codec_process() start");
+
+	if (cd->codec.state < CODEC_PREPARED) {
+		comp_err(dev, "codec_prepare(): wrong state of codec %x, state %d",
+			 cd->ca_config.codec_id, codec->state);
+		return -EPERM;
+	}
+
+	ret = codec->ops->process(dev);
+	if (ret) {
+		comp_err(dev, "codec_prepare() error %d: codec process failed for codec_id %x",
+			 ret, codec_id);
+		goto out;
+	}
+
+	comp_dbg(dev, "codec_process() done");
+out:
+	return ret;
+}
+
+int codec_apply_runtime_config(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint32_t codec_id = cd->ca_config.codec_id;
+	struct codec_data *codec = &cd->codec;
+
+	comp_dbg(dev, "codec_apply_config() start");
+
+	if (cd->codec.state < CODEC_PREPARED)
+		return -EPERM;
+
+	if (cd->codec.state < CODEC_PREPARED) {
+		comp_err(dev, "codec_prepare() wrong state of codec %x, state %d",
+			 cd->ca_config.codec_id, codec->state);
+		return -EPERM;
+	}
+
+	ret = codec->ops->apply_config(dev);
+	if (ret) {
+		comp_err(dev, "codec_apply_config() error %d: codec process failed for codec_id %x",
+			 ret, codec_id);
+		goto out;
+	}
+
+	cd->codec.r_cfg.avail = false;
+	rfree(codec->r_cfg.data);
+	codec->r_cfg.data = NULL;
+
+	comp_dbg(dev, "codec_apply_config() end");
+out:
+	return ret;
+}
+
+int codec_reset(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct codec_data *codec = &cd->codec;
+
+	ret = codec->ops->reset(dev);
+	if (ret) {
+		comp_err(dev, "codec_apply_config() error %d: codec specific .reset() failed for codec_id %x",
+			 ret, cd->ca_config.codec_id);
+		return ret;
+	}
+
+	codec->r_cfg.avail = false;
+	codec->r_cfg.size = 0;
+	rfree(codec->r_cfg.data);
+
+	codec->state = CODEC_PREPARED;
```

Comment:
Are we really prepared after reset ? Shouldn't prepare need to be called ?

Reply:
Yes the idea is to just reset codec to its initial state - reset runtime configuration like volume up/down, certain frequencies damping etc, but the memory allocated for the codec should remain intact.

Reply:
ok, best to have a comment here describing.